pkg/activator/handler/concurrency_reporter.go

```diff
@@ -61,7 +73,104 @@ func NewConcurrencyReporter(ctx context.Context, podName string,
 		reqCh:   reqCh,
 		statCh:  statCh,
 		rl:      revisioninformer.Get(ctx).Lister(),
+
+		stats:                make(map[types.NamespacedName]*network.RequestStats),
+		reportedFirstRequest: make(map[types.NamespacedName]float64),
+	}
+}
+
+// handleEvent handles request events (in, out) and updates the respective stats.
+func (cr *ConcurrencyReporter) handleEvent(event network.ReqEvent) {
+	stats, msgs := cr.getOrCreateStat(event)
+	if len(msgs) > 0 {
+		cr.statCh <- msgs
+	}
+	stats.HandleEvent(event)
+}
+
+// getOrCreateStat gets a stat from the state if present.
+// If absent it creates a new one and returns it, potentially returning a StatMessage too
+// to trigger an immediate scale-from-0.
+func (cr *ConcurrencyReporter) getOrCreateStat(event network.ReqEvent) (*network.RequestStats, []asmetrics.StatMessage) {
+	cr.mux.RLock()
+	stat := cr.stats[event.Key]
+	cr.mux.RUnlock()
+
+	if stat != nil {
+		return stat, nil
+	}
+
+	// Doubly checked locking.
+	cr.mux.Lock()
+	defer cr.mux.Unlock()
+
+	stat = cr.stats[event.Key]
+	if stat != nil {
+		return stat, nil
+	}
+
+	stat = network.NewRequestStats(event.Time)
+	cr.stats[event.Key] = stat
+
+	if event.Type == network.ReqIn {
+		// Return a StatMessage if this is an incoming request.
```

Comment:
should this be an error (log, panic, etc) if this is a reqout?

Reply:
Hmm, maybe we should just drop the case. I don't think it can happen at all.

Reply:
it should not, no.